proposals/2023-08-21-utf8.md

```diff
@@ -0,0 +1,159 @@
+* Add a way to escape metric names for clients that do not support the full character set.
+* Update endpoint code to support html-unquoting metric names in http API calls.
+
+### prometheus/common
+
+We will add a boolean argument to model.IsValidMetricName that switches validation logic to allow UTF8 names (and therefore will only check for 0 length). Because this is a public function, all users of this library will need to add an argument to their call sites.
+
+openmetrics_create.go and text_create.go will detect if the metric names require quoting, and if so, use 
+the new syntax when creating the formatted text.
+
+Other repositories that use this check to filter out "bad" metric names will have to be optional via a flag or argument.
+
+### prometheus/client_golang
+
+The golang library checks that metrics have a valid name when NewDesc is called. We can create a new parallel function, NewDescUTF8 (or similar) that will bypass the validity check.
+
+### prometheus/prometheus
+
+All support for UTF8 will be behind a default-off configuration flag. Behavior will not be changed unless this flag is on.
+
+Prometheus itself needs updates to its lexers and parsers to support reading the new exposition syntax and promql syntax:
+
+* openmetricslex
+* promlex
+* generated_parser
+
+#### Backwards-compatibility
+
+Prometheus will also need to support queries and responses from old clients that do not support the new syntax. We will create a new boolean HTTP header called AllowUTF8Names. When this value is true in the request, and the feature is enabled in prometheus, UTF8 support will be enabled. For any client that does not pass the AllowUTF8Names header, any HTTP request that can return metric names (including simple queries) will have the UTF8 metric names escaped. 
+
+#### Text escaping
+
+We propose an escaping syntax for metric names that contain UTF8 characters:
+
+* Prefix the string with `U__`
+* All UTF8 characters will be encoded as underscores surrounding the unicode value, like `_1F60A_`
+* All pre-existing underscores will become doubled: `__`
+* If a string should start with "U__" already, it will need to be escaped: `U___55_____`. (That's `U__` + `_55_` + `__` + `__`)
+
+#### HTTP escaping
+
+The HTTP endpoints will need to support html unescaping metric names where they occur.
+
+## Alternatives
+
+### Non-quoted metric names
+
+`my.dotted.metric{label="value"}`
+
+While this syntax could be made to work with simple additional characters like dots, it does not fulfill the goal of full UTF8 support, which would include `{` or other special characters without a verbose and unreadable escaping mechanism. Selecting this approach would introduce a convenient syntax for only a small new subset of UTF8, so a quoting mechanism would still be needed for the full character set. Rather than implement this half-improvement we decided to only support a full quoting approach.
```

Comment:
Could you add an example of how the escaping would have to work for special characters?

Reply:
done.  I also fixed some of the syntax examples to match